### DISCUSSION TITLE
How to Use ApiResponse Without a Return Type?

### DISCUSSION THREAD

**Discussion by sdepouw** [reactiveui/refit#1594] [Q&A] [answered]
Hello. I have a simple scenario: Hit an endpoint that returns nothing, but be able to inspect the response code/etc.

If I didn't care about the response I'd do something like:

```csharp
[Post("/foo")]
public Task Foo(RequestBody request, CancellationToken cancellationToken);
```

And if the HTTP hit returned some response, I'd model it with `ApiResponse<T>`:

```csharp
[Post("/foo")]
public Task<ApiResponse<ResponseDTO>> Foo(RequestBody request, CancellationToken cancellationToken);
```

However, what if there is no response? `ApiResponse` (without type) is internal and cannot be used. I suppose technically there's always some content response (in the form of an empty string), but I think it'd be silly to have to define my endpoint/contract as "this returns a string" when one is never returned. So for now, my workaround is this:

```csharp
[Post("/foo")]
public Task<ApiResponse<string>> Foo(RequestBody request, CancellationToken cancellationToken); // the string is always empty, but users of this client won't know that!
```

Is there a way to use `ApiResponse` so I can check the status code, ensure success, etc. without having to provide a type?

**Comment by glennawatson** [comment, marked as answer]
Use the non-generic `IApiResponse` as the return type. It is public and gives you status code, headers, success flags and error without forcing a content type.

```csharp
[Post("/foo")]
Task<IApiResponse> Foo(RequestBody request, CancellationToken cancellationToken);
```

Then:

```csharp
using var response = await api.Foo(request, ct);
if (response.IsSuccessStatusCode) { ... }
var code = response.StatusCode;
```

Notes:
- `IApiResponse` (no type parameter) is the public, content-less response wrapper. It is verified as a supported return type (see ResponseTests.cs which uses `Task<IApiResponse>` and `ValueTask<IApiResponse>`).
- Dispose it (or use `using`) since it owns the underlying response.
- This avoids the `ApiResponse<string>` workaround entirely.
